papers/oleksandr_pavlyk/optimizations.rst

```diff
@@ -72,9 +72,129 @@ The most significant improvements in memory optimizations in Update 2 comes from
 
 Faster Machine Learning with Scikit-learn
 -----------------------------------------
-Scikit-learn is among the most popular Python machine learning packages. The initial release of Intel Distribution for Python provided Scikit-learn optimizations via respective NumPy and SciPy functions accelerated by Intel MKL. Update 2 optimizes selective key machine learning algorithms in Scikit-learn, accelerating them with the Intel |R| Data Analytics Acceleration Library (Intel |R| DAAL).
+Scikit-learn is well know library that provides a log of algorithms for many areas of machine learning.
+Having limited developer resources this project prefers universal solutions and proven algorithms.
+For performance improvement scikit-learn uses Cython and underling (through SciPy and Numpy) BLAS/LAPACK libraries.
+OpenBLAS and MKL uses threaded based parallelism to utilize multicores of modern CPUs.
+Unfortunately  BLAS/LAPACK’s functions are too low level primitives and their usage is often not very efficient comparing to possible high-level parallelism.
+For high-level parallelism scikit-learn uses multiprocessing approach that is not very efficient from technical point of view.
+On the other hand Intel provides Intel |R| Data Analytics Acceleration Library (Intel(R) DAAL) that helps speed up big data analysis by providing highly optimized algorithmic building blocks for all stages of data analytics (preprocessing, transformation, analysis, modeling, validation, and decision making) in batch, online, and distributed processing modes of computation.
+It is originally written in C++ and provides Java and Python bindings.
+In spite of the fact that DAAL is heavily optimized for all Intel Architectures including Xeon Phi it’s very questionable how to use DAAL binding from Python.
+DAAL bindings for python are generated automatically and reflects original C++ API very close that makes its usage quite complicated because of not native for python programmers idioms and chary documentation.
+
+To mix the power of well optimized for HW native code with familiar ML API Intel Python Distribuition started efforts on scikit-learn optimization.
+Therefore, beginning from 2017 U2 IDP includes scikit-learn with daal4sklearn module.
 Specifically, Update 2 optimizes Principal Component Analysis (PCA), Linear and Ridge Regressions, Correlation and Cosine Distances, and K-Means. Speedups may range from 1.5x to 160x.
 
+There are no exact matching between sklearn and DAAL API and they aren’t fully compatible for all inputs so for the cases when daal4sklearn detects incompatibility it fallbacks to original sklearn’s implementation.
```

Comment:
It should " ... it falls back on ...", see [TFD:fall back on](http://www.tfd.com/fall+back+on).